### PR TITLE
Fix fall-through warnings

### PR DESCRIPTION
**Pull request by lucasdemarchi**
GCC 7 turned the fallthrough warning on by default. These are the ways
to handle them on the cases it should fallthrough:

```c++
    switch (cond) {
    case 1:
        bar (1);
        __attribute__ ((fallthrough)); // C and C++03
    case 2:
        bar (2);
        [[gnu::fallthrough]]; // C++11 and C++14
    case 3:
        bar (3);
        [[fallthrough]]; // C++17 and above
    /* ... */
    }
```

More info: https://dzone.com/articles/implicit-fallthrough-in-gcc-7
And it **does** work with clang as well.

### CODE REVIEW

**Comment by WickedShell** [comment]
The thought on the devcall when I brought this up (around a month ago) but I never sent in a PR was to address it with the use of // fallthrough comments and `-Wimplicit-fallthrough=2` rather then using any compilier/C++ version attributes.

I know I ran into an issue with one of the ones you outlined above when I tried to use it with GCC 4.9, but I can't recall which it was.

Either way I'm happy to see something improving this, as my build warnings are currently insane to the point where I know I miss warnings that I do want to fix.

**Comment by lucasdemarchi** [comment]
I couldn't get any of the options with comments to work on GCC 7.  This one I'm sending works on GCC (tested on at least 4.9 but probably olders as well) and clang.

**Comment by WickedShell** [comment]
This passes in GCC 7.1.1 with ```-Wimplicit-fallthrough=3```
```
int main(int argc, char **argv) {                                                                                                                        
        int a = 0;

switch (argc)
   {   
   case 1:
     a = 1;
     // fallthrough
   case 2:
     a = 2;
     break;
   /* ... */
   }   
  return a;
}
```

I assume the main difficulty is getting the flag passed down to everywhere it's needed?

**Comment by lucasdemarchi** [comment]
Humn... I could get it to work in C but not in C++ when I tried with comments. But after checking these attributes I figured it's actually better to instruct the compiler about the fallthrough than to rely on comment parsing. What's the advantage of having it as comment?

**Comment by WickedShell** [comment]
My concern is I get these
```
test.c:8:6: warning: attributes at the beginning of statement are ignored [-Wattributes]
      [[gnu::fallthrough]];
```

However that seems to be what your -Wno-attributes patch fixes. I guess I prefer the fact that I can just add `// fallthrough` without having to mess with the flags at all, and it just works for old code and it doesn't imply any other flags for older compilers. However I hadn't found the -Wno-attributes flag when I did my original testing so if you'd prefer the attribute one and it doesn't create more warnings I'm fine with it.

My other concern is that this doesn't work with clang, as clang requires `[[clang::fallthrough]]`

What if we took this approach here: http://clang.llvm.org/docs/LanguageExtensions.html#has-cpp-attribute and use that to create a define that is used? This would give us the warning with clang and GCC as needed, and could be a no-op for older GCC versions without needing to mess with the compiler flags at all?

**Comment by lucasdemarchi** [comment]
This came from mavlink-router which previously was using the has_attribute approach. It was replaced after I found out this also works with clang. You have to pass -std=gnu++11 as we already do for other reasons.
https://github.com/01org/mavlink-router/commit/1a20d012f62f64da7c183f6dba792caf6bf63962

Minimum required version for GCC is 4.9 and it does work with that, so I don't see why we would care about ancient compilers.

**Comment by WickedShell** [comment]
I'd maintain that clang doesn't accept the `[[gnu::fallthrough]]` attribute, but it does accept `[[fallthrough]]` (at least on clang version 4.0.1 (tags/RELEASE_401/final) ) so if we could at least stick with that version of the attribute I'd be happier. (And that does worth with GCC 7.1 correctly)

**Comment by lucasdemarchi** [comment]
@WickedShell I updated it to use [[fallthrough]]

**Comment by tridge** [comment]
maybe use a #define FALLTHROUGH which checks compiler version and chooses the right one?
like we do for PACKED

**Comment by lucasdemarchi** [comment]
Do we need to? If it works with the compilers we use, I don't see a compelling reason.

**Comment by lucasdemarchi** [comment]
> maybe use a #define FALLTHROUGH which checks compiler version and chooses the right one?
> like we do for PACKED

We actually don't check what's the correct one on that case because there's only one that's right nowadays (there could be a pragma, but that's out of scope here).

With the intention to get this applied sooner than later I added the define and will apply once CI passes...  the current warnings are annoying and getting this type pf PR up-to-date is a pain since it touches everywhere.

**Comment by tridge** [comment]
> Do we need to? If it works with the compilers we use, I don't see a compelling reason.

it was just a suggestion, I didn't intend to block the PR on it. It just means if we do hit a problem with some compiler then it's very easy to fix in one place.

**Comment by lucasdemarchi** [comment]
> it was just a suggestion, I didn't intend to block the PR on it. It just means if we do hit a problem with some compiler then it's very easy to fix in one place.

already changed now :). I don't like adding a macro like that, but I don't care much.

File: libraries/AP_GPS/AP_GPS_SIRF.cpp [hunk 1]
```diff
@@ -204,48 +204,48 @@ AP_GPS_SIRF::_accumulate(uint8_t val)
 /*
   detect a SIRF GPS
  */
-bool
-AP_GPS_SIRF::_detect(struct SIRF_detect_state &state, uint8_t data)
+bool AP_GPS_SIRF::_detect(struct SIRF_detect_state &state, uint8_t data)
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_SIRF.cpp]: Whitespace changes, but they look to be correct, and not worth the effort of rebasing again IMO.

Comment (lucasdemarchi) [inline libraries/AP_GPS/AP_GPS_SIRF.cpp]: They were done on a separate commit, which is the right policy for these fixes :)

Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_SIRF.cpp]: /me whistles

I only peaked at the overall diff in github, sorry :)
